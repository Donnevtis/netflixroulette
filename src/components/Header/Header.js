import React, { Component, createRef } from 'react';
import { func, string } from 'prop-types';
import styles from './Header.module.scss';

class Header extends Component {
  inputRef = createRef();

  handleSubmit = (event) => {
    event.preventDefault();

    const { onSubmit, query } = this.props;
    const newQuery = this.inputRef.current.value;

    if (newQuery !== query) {
      onSubmit({ query: newQuery });
    }
  };

  render() {
    const { onOpenAdd } = this.props;

    return (
      <header className={styles.header}>
        <div className={styles.topWrapper}>
          <span className={styles.title}>
            <b>netflix</b>
            roulette
          </span>
          <button type="button" className={styles.addMovieBtn} onClick={onOpenAdd}>
            + ADD MOVIE
          </button>
        </div>
        <form className={styles.searchForm} onSubmit={this.handleSubmit}>
          <fieldset className={styles.searchForm__fieldset}>
            <legend className={styles.searchForm__label}>FIND YOUR MOVIE</legend>
            <input
              type="text"
              className={styles.searchInput}
              placeholder="What do you want to watch?"
              onChange={this.handleChange}
              ref={this.inputRef}
            />
            <button type="submit" className={styles.searchBtn}>
              SEARCH
            </button>
          </fieldset>
        </form>
      </header>
    );
  }
}

Header.propTypes = {
  onOpenAdd: func.isRequired,
  onSubmit: func.isRequired,
  query: string.isRequired,
};

export default Header;
