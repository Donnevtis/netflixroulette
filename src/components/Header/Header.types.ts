import { Dispatch, SetStateAction } from 'react';

export type HeaderProps = {
  onOpenAdd: () => void;
  onSubmit: Dispatch<SetStateAction<string>>;
  query: string;
};
