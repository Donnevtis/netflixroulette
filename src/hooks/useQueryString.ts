import { useSearchParams } from 'react-router-dom';

const useQueryString = () => {
  const [searchParams, setSearchParams] = useSearchParams();

  return (key: string, value: string) => {
    searchParams.set(key, value);
    setSearchParams(searchParams);
  };
};

export default useQueryString;
