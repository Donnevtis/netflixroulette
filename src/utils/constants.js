const GENRES = [
  'Action',
  'Adventure',
  'Animation',
  'Comedy',
  'Crime',
  'Documentary',
  'Drama',
  'Family',
  'Fantasy',
  'Horror',
  'Romance',
  'Science Fiction',
];

const NAV_GENRES = ['All', 'Documentary', 'Comedy', 'Horror', 'Crime'];

const SORTINGMAP = [
  ['Release date', 'release_date'],
  ['Title', 'title'],
  ['Raiting', 'vote_average'],
  ['Popularity', 'vote_count'],
  ['Duration', 'runtime'],
];

const SORT_BY = SORTINGMAP.map((opt) => opt[0]);

const IMG_PLACEHOLDER = 'https://via.placeholder.com/500x750?text=Image+not+found';

export {
  GENRES, NAV_GENRES, SORTINGMAP, SORT_BY, IMG_PLACEHOLDER,
};
