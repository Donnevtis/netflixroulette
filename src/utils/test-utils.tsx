/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { Provider } from 'react-redux';
import store from '@src/store';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

export const renderWithStore = (ui: JSX.Element, options?: Parameters<typeof render>[1]) => ({
  ...render(<Provider store={store}>{ui}</Provider>, options),
  store,
});

export const renderWithRouter = (ui: JSX.Element, options?: Parameters<typeof render>[1]) =>
  render(<MemoryRouter>{ui}</MemoryRouter>, options);

export const renderWithRouterAndStore = (
  ui: JSX.Element,
  options: Parameters<typeof render>[1] = {},
) => renderWithStore(ui, { ...options, wrapper: MemoryRouter });
