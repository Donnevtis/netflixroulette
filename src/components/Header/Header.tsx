import React, { SyntheticEvent, useContext, useRef } from 'react';
import AppContext from '@src/context/app.context';
import { RootState } from '@src/types';
import { connect } from 'react-redux';
import { fetchMovies } from '@src/store/actionCreators/movies';
import { selectQuery } from '@src/store/selectors/movies.selectors';
import { HeaderProps } from './Header.types';
import styles from './Header.module.scss';

import Title from '../Title/Title';

const Header = ({ filterMovies, query }: HeaderProps) => {
  const { setShowAdd } = useContext(AppContext);
  const inputRef = useRef<HTMLInputElement>(null);

  const handleClick = () => setShowAdd(true);

  const handleSubmit = (event: SyntheticEvent) => {
    event.preventDefault();
    const { current } = inputRef;

    if (!current) {
      return;
    }

    const newQuery = current.value;

    if (newQuery !== query) {
      filterMovies({ query: newQuery });
    }
  };

  return (
    <header className={styles.header}>
      <div className={styles.topWrapper}>
        <Title />
        <button type="button" className={styles.addMovieBtn} onClick={handleClick}>
          + ADD MOVIE
        </button>
      </div>
      <form className={styles.searchForm} onSubmit={handleSubmit}>
        <fieldset className={styles.searchForm__fieldset}>
          <legend className={styles.searchForm__label}>FIND YOUR MOVIE</legend>
          <input
            type="text"
            className={styles.searchInput}
            placeholder="What do you want to watch?"
            ref={inputRef}
            defaultValue={query}
          />
          <button type="submit" className={styles.searchBtn}>
            SEARCH
          </button>
        </fieldset>
      </form>
    </header>
  );
};

const mapStateToProps = (state: RootState) => ({ query: selectQuery(state) });

const mapDispatchToProps = {
  filterMovies: fetchMovies,
};

export default connect(mapStateToProps, mapDispatchToProps)(Header);
