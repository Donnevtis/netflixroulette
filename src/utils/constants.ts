import { isEmpty, notSelected, isNumber, lessThan, greaterThan } from '@src/utils/validators';
import { Genre, GenreFilters, SortFilters, SortBy, Movie, MovieDraft } from '@src/types';

export const GENRES = Object.keys(Genre) as Genre[];

export const NAV_GENRES = Object.keys(GenreFilters) as GenreFilters[];

export const SORT_BY = Object.keys(SortFilters) as SortBy[];

export const IMG_PLACEHOLDER = 'https://via.placeholder.com/500x750?text=Image+not+found';

export const DEFAULT_MOVIE: Movie = {
  id: 0,
  title: '',
  tagline: '',
  vote_average: 0,
  vote_count: 0,
  release_date: '',
  poster_path: '',
  overview: '',
  budget: 0,
  revenue: 0,
  genres: [],
  runtime: 0,
};

export const VALIDATORS_SCHEME = {
  title: [isEmpty],
  poster_path: [isEmpty],
  genres: [notSelected],
  release_date: [isEmpty],
  vote_average: [isNumber, lessThan(10), greaterThan(0), isEmpty],
  runtime: [isNumber, greaterThan(0), isEmpty],
  overview: [isEmpty],
};

export const FIELDS = Object.keys(VALIDATORS_SCHEME) as (keyof MovieDraft)[];

export const RIGHT_OFFSET = 210;
export const BOTTOM_OFFSET = 115;
