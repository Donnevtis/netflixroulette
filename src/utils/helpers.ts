import { Genre, GenreRecord, MovieDraft } from '@src/types';
import { GENRES, VALIDATORS_SCHEME } from './constants';

export const arrayToObject = (array: string[] = []): GenreRecord => {
  const genres: { [key: string]: boolean } = {};

  GENRES.forEach((genre) => {
    genres[genre] = array.includes(genre);
  });

  return genres as GenreRecord;
};

export const objectToArray = (
  genres: GenreRecord,
): Genre[] => GENRES.filter((genre) => genres[genre]);

export const numToString = (num: number): string => (num ? String(num) : '');

export const capitalize = (str: string): string => str.replace(/^\w/, (s: string) => s.toUpperCase());

export const extractYear = (
  stringDate: string,
): string => String(new Date(stringDate).getFullYear());

export const minutesToHours = (minutes: number) => {
  const hours = Math.floor(minutes / 60);
  const rest = Math.floor(minutes % 60);

  return [hours && `${hours}h`, rest && `${rest}min`]
    .filter((_) => _)
    .join(' ');
};

export const validate = (
  fieldName: keyof MovieDraft,
  value: string | GenreRecord,
) => {
  let errorMessage = '';

  VALIDATORS_SCHEME[fieldName].forEach(({ test, error }) => {
    if (test(value)) {
      errorMessage = error;
    }
  });

  return errorMessage;
};
