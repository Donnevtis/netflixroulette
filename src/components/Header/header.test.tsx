import React from 'react';
import { screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { renderWithRouterAndStore } from '@src/utils/test-utils';
import useQueryString from '@src/hooks/useQueryString';
import Header from './Header';

jest.mock('@src/hooks/useQueryString');
const mockedSetQueryParams = jest.spyOn({ useQueryString }, 'useQueryString');

describe('Header', () => {
  test('should render a search panel', async () => {
    const { container } = renderWithRouterAndStore(<Header />);
    expect(container).toMatchSnapshot();
  });

  test('changes a query parameter when the search button is clicked', async () => {
    const setQueryParams = jest.fn();
    mockedSetQueryParams.mockReturnValue(setQueryParams);

    renderWithRouterAndStore(<Header />);

    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: 'matrix' },
    });

    expect(screen.getByRole('textbox')).toHaveValue('matrix');
    expect(setQueryParams).not.toHaveBeenCalled();

    fireEvent.click(screen.getByRole('button', { name: /search/i }));

    expect(setQueryParams).toHaveBeenCalledWith({ query: 'matrix' });
  });
});
