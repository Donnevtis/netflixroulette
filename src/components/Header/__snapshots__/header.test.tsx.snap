// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render a search panel 1`] = `
<div>
  <header>
    <div>
      <span>
        <b>
          netflix
        </b>
        roulette
      </span>
      <button
        type="button"
      >
        + ADD MOVIE
      </button>
    </div>
    <form>
      <fieldset>
        <legend>
          FIND YOUR MOVIE
        </legend>
        <input
          placeholder="What do you want to watch?"
          type="text"
          value=""
        />
        <button
          type="submit"
        >
          SEARCH
        </button>
      </fieldset>
    </form>
  </header>
</div>
`;
